File: 2pac/2.01-Heap/Heap.js
function Heap(){
    this.build = HeapBuild;
    this.printHeap = HeapPrintHeap;
}
    function HeapBuild(list){
        n = list.getLength();
        h = Math.ceil( Math.log2( n + 1) );
        mi = Math.pow(2, h-1) - 1;
        mf = mi * 2 ;
        w = mf + 1;

        console.log(list)
        console.log("n: " + n )
        console.log("h: " + h )
        console.log("mi: " + mi )
        console.log("mf: " + mf )
        console.log("w: " + w )
        
        matrix = [];
        for(var i=0;i<h;i++){
            row = [];
            for(var j=0;j<w;j++){
                row.push("&nbsp&nbsp&nbsp&nbsp");
            }
            matrix.push(row);
        }
        
        half = Math.floor(w/2);
        count = 0;
        matrix[0][half] = list.atPosition(count).value; //Primera fila
        count++;
        
        for(var i=1;i<h;i++){
            for(var j=0;j<w;j++){
                if(j == 0){
                    bp = Math.pow(2, h-i) - 1 ; //BlankSpace
                    bsi = Math.floor(bp/2); //BlankSpaceInit
                    j += bsi;
                    matrix[i][j] = list.atPosition(count).value;
                    count++;
                }
                else if(j + bp < w && count<n){
                    j += bp;
                    matrix[i][j] = list.atPosition(count).value;
                    count++;
                }
            }
        }
        this.printHeap(matrix);
    }

    function HeapPrintHeap(myArray = []){
        stringHtml = "<table border='5'>";

        for(i=0;i<myArray.length;i++){
            stringHtml += `<tr><td>${myArray[i].join("</td><td>")}</td></tr>`;
        }
        stringHtml += "</table>";

        document.write(stringHtml);
    }

File: Extras/Repaso/Heap.py
from BST import LinkedList
import math
class Heap:
    def __init__(self):
        pass
    def build(self,LL):
        n = LL.getLength()
        h = math.ceil( math.log2( n + 1) )
        mi = int(math.pow(2, h-1) - 1)
        mf = mi * 2 
        w = mf + 1
    
        matriz = [["&nbsp" for i in range(0,w)] for j in range(0,h)]
        half = math.floor(w/2)
        count = 0
        matriz[0][half] = LL.atPosition(count)
        count += 1
        for i in range(1,h):
            bs = int (math.pow(2,h-i) - 1)
            bsi = math.floor( bs/2 )
            j=0
            while(j<w):
                if(j==0):
                    j += bsi
                    matriz[i][j] = LL.atPosition(count)
                    count += 1
                elif(j+bs < w and count<=n):
                    j += bs
                    matriz[i][j] = LL.atPosition(count)
                    count += 1
                j+=1
        self.MakeHTMLHeap(matriz)

    def MakeHTMLHeap(self, matrix):
        f = open("Extras//Repaso//indexHeap.html","w")
        htmlStr = self.htmlTable(matrix)
        f.write(htmlStr)
        f.close()

    def htmlTable(self,matrix):
        htmlArr = ["<table border='5'>"]
        for r in matrix:
            row = []
            for value in r:
                row.append(str(value))
            htmlArr.append("<tr><td>")
            htmlArr.append("</td><td>".join(row))
            htmlArr.append("</td></tr>")
        htmlArr.append("</table>")
        return "".join(htmlArr)


'''
h = Heap()
l = LinkedList()
l.add(5)
l.add(2)
l.add(7)
l.add(1)
l.add(4)
l.add(6)
l.add(8)
h.build(l)
'''